File: server/utils/jobs.ts
// Cola de jobs secuencial (una GPU) + eventos para SSE. Port de core/jobs.py
// al modelo del event loop de Node: cada job ejecuta un worker (subproceso Python
// en real, o simulacion en Node); sin hilos.
import { basename, join } from 'pathe'
import { jobsDir } from './config'
import { sha1Hex } from './hash'
import { runJob } from './worker'

export interface JobProgress {
  stage?: string
  frames_done?: number
  frames_total?: number
  fps?: number
  eta_s?: number
  chunk?: number
  chunks_total?: number
  thumbnail?: string
}

export interface Job {
  id: string
  kind: 'demo' | 'full'
  source: string
  filename: string
  cfg: Record<string, any>
  probe: Record<string, any>
  segment_start: number
  segment_duration: number
  state: 'queued' | 'running' | 'done' | 'error' | 'cancelled' | 'paused'
  created_at: number
  progress: JobProgress
  output: string | null
  error: string | null
  simulate: boolean
}

type Listener = (payload: string) => void

// Serializacion canonica y estable (claves ordenadas) para el hash del workdir.
function canonical(value: any): string {
  if (value === null || typeof value !== 'object') return JSON.stringify(value)
  if (Array.isArray(value)) return '[' + value.map(canonical).join(',') + ']'
  const keys = Object.keys(value).sort()
  return '{' + keys.map(k => JSON.stringify(k) + ':' + canonical(value[k])).join(',') + '}'
}

class JobManager {
  jobs = new Map<string, Job>()
  private queue: string[] = []
  private running = false
  private controllers = new Map<string, AbortController>()
  private listeners = new Set<Listener>()
  private eventId = 0

  submit(kind: 'demo' | 'full', source: string, cfg: Record<string, any>,
    probe: Record<string, any>, segmentStart: number, segmentDuration: number,
    simulate: boolean): Job {
    const id = crypto.randomUUID().replace(/-/g, '').slice(0, 12)
    const job: Job = {
      id, kind, source, filename: basename(source), cfg, probe,
      segment_start: segmentStart, segment_duration: segmentDuration,
      state: 'queued', created_at: Date.now() / 1000, progress: {},
      output: null, error: null, simulate,
    }
    this.jobs.set(id, job)
    this.controllers.set(id, new AbortController())
    this.queue.push(id)
    this.emit(job)
    void this.pump()
    return job
  }

  cancel(id: string): boolean {
    const job = this.jobs.get(id)
    if (!job) return false
    this.controllers.get(id)?.abort()
    if (job.state === 'queued') {
      job.state = 'cancelled'
      this.queue = this.queue.filter(x => x !== id)
      this.emit(job)
    }
    return true
  }

  listJobs(): Job[] {
    return [...this.jobs.values()].sort((a, b) => a.created_at - b.created_at)
  }

  // Directorio determinista por (fichero, cfg, kind, segmento) — reanudacion.
  // Calculo PURO (sin efectos): el mkdir lo hace worker.runReal justo antes de
  // escribir el jobspec; los lectores (preview) no necesitan crear nada.
  async workdirPath(job: Job): Promise<string> {
    const key = canonical([job.source, job.cfg, job.kind, job.segment_start, job.segment_duration])
    const digest = (await sha1Hex(key)).slice(0, 16)
    return join(jobsDir(), digest)
  }

  publicJob(job: Job): Omit<Job, 'probe'> {
    const { probe, ...pub } = job
    return pub
  }

  subscribe(cb: Listener): () => void {
    this.listeners.add(cb)
    return () => this.listeners.delete(cb)
  }

  private emit(job: Job): void {
    this.eventId++
    const payload = JSON.stringify({ id: this.eventId, job: this.publicJob(job) })
    for (const cb of [...this.listeners]) {
      try { cb(payload) } catch { /* suscriptor caido */ }
    }
  }

  private async pump(): Promise<void> {
    if (this.running) return
    this.running = true
    try {
      while (this.queue.length) {
        const id = this.queue.shift()!
        const job = this.jobs.get(id)
        if (!job || job.state === 'cancelled') continue
        const ctrl = this.controllers.get(id)!
        job.state = 'running'
        this.emit(job)

        let last = 0
        const onProgress = (p: JobProgress) => {
          job.progress = { ...job.progress, ...p }
          const now = Date.now()
          if (now - last > 250 || p.stage === 'completado') { last = now; this.emit(job) }
        }

        try {
          const out = await runJob(job, await this.workdirPath(job), onProgress, ctrl.signal)
          if (ctrl.signal.aborted) job.state = 'cancelled'
          else { job.output = out; job.state = 'done' }
        } catch (e: any) {
          if (ctrl.signal.aborted) job.state = 'cancelled'
          else { job.state = 'error'; job.error = `${e?.name || 'Error'}: ${e?.message ?? e}` }
        }
        this.emit(job)
      }
    } finally {
      this.running = false
    }
  }
}

export const manager = new JobManager()

File: worker/backend/core/jobs.py
"""Gestor de trabajos: cola secuencial (una GPU), eventos SSE, cancelación."""
from __future__ import annotations

import asyncio
import json
import queue as _queue
import threading
import time
import uuid
from dataclasses import dataclass, field, asdict
from pathlib import Path

from backend import config
from backend.pipeline.runner import RunnerContext, Progress, CancelledError


@dataclass
class Job:
    id: str
    kind: str                  # 'demo' | 'full'
    source: str
    filename: str
    cfg: dict
    probe: dict
    segment_start: float
    segment_duration: float
    state: str = "queued"      # queued|running|done|error|cancelled
    created_at: float = field(default_factory=time.time)
    progress: dict = field(default_factory=dict)
    output: str | None = None
    error: str | None = None
    simulate: bool = False

    def public(self) -> dict:
        d = asdict(self)
        d.pop("probe", None)
        return d


class JobManager:
    def __init__(self):
        self.jobs: dict[str, Job] = {}
        self._cancel_events: dict[str, threading.Event] = {}
        self._queue: _queue.Queue[str] = _queue.Queue()
        self._subscribers: list[asyncio.Queue] = []
        self._loop: asyncio.AbstractEventLoop | None = None
        self._event_id = 0
        self._worker = threading.Thread(target=self._work_loop, daemon=True,
                                        name="job-worker")
        self._worker.start()

    def attach_loop(self, loop: asyncio.AbstractEventLoop) -> None:
        self._loop = loop

    # ── API pública ─────────────────────────────────────────────────────────
    def submit(self, kind: str, source: str, cfg: dict, probe: dict,
               segment_start: float, segment_duration: float,
               simulate: bool) -> Job:
        job = Job(id=uuid.uuid4().hex[:12], kind=kind, source=source,
                  filename=Path(source).name, cfg=cfg, probe=probe,
                  segment_start=segment_start,
                  segment_duration=segment_duration, simulate=simulate)
        self.jobs[job.id] = job
        self._cancel_events[job.id] = threading.Event()
        self._queue.put(job.id)
        self._emit(job)
        return job

    def cancel(self, job_id: str) -> bool:
        job = self.jobs.get(job_id)
        if not job:
            return False
        ev = self._cancel_events.get(job_id)
        if ev:
            ev.set()
        if job.state == "queued":
            job.state = "cancelled"
            self._emit(job)
        return True

    def workdir(self, job: Job) -> Path:
        """Directorio determinista por (fichero, config, segmento) — NO por id:
        así relanzar el mismo trabajo tras reiniciar la app reutiliza los
        chunks ya completados (reanudación real)."""
        import hashlib
        import json as _json
        key = _json.dumps([job.source, job.cfg, job.kind, job.segment_start,
                           job.segment_duration], sort_keys=True)
        digest = hashlib.sha1(key.encode()).hexdigest()[:16]
        d = config.JOBS_DIR / digest
        d.mkdir(parents=True, exist_ok=True)
        return d

    def subscribe(self) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue(maxsize=500)
        self._subscribers.append(q)
        return q

    def unsubscribe(self, q: asyncio.Queue) -> None:
        if q in self._subscribers:
            self._subscribers.remove(q)

    # ── interno ──────────────────────────────────────────────────────────────
    def _emit(self, job: Job) -> None:
        if self._loop is None:
            return
        self._event_id += 1
        payload = json.dumps({"id": self._event_id, "job": job.public()},
                             ensure_ascii=False)
        for q in list(self._subscribers):
            try:
                self._loop.call_soon_threadsafe(q.put_nowait, payload)
            except RuntimeError:
                pass

    def _work_loop(self) -> None:
        while True:
            job_id = self._queue.get()
            job = self.jobs.get(job_id)
            if job is None or job.state == "cancelled":
                continue
            cancel = self._cancel_events[job_id]
            job.state = "running"
            self._emit(job)

            last_emit = 0.0

            def on_progress(p: Progress, _job=job):
                nonlocal last_emit
                _job.progress = {k: v for k, v in asdict(p).items() if v is not None}
                now = time.monotonic()
                # limitar a ~4 eventos/s salvo cambios de etapa
                if now - last_emit > 0.25 or p.stage in ("completado",):
                    last_emit = now
                    self._emit(_job)

            try:
                ctx = RunnerContext(
                    job_id=job.id, source=job.source, probe=job.probe,
                    cfg=job.cfg, segment_start=job.segment_start,
                    segment_duration=job.segment_duration,
                    is_demo=(job.kind == "demo"), workdir=self.workdir(job),
                    on_progress=on_progress, cancel=cancel,
                    simulate=job.simulate)
                out = ctx.run()
                job.output = str(out)
                job.state = "done"
            except CancelledError:
                job.state = "cancelled"
            except Exception as e:  # noqa: BLE001 — el error viaja a la UI
                job.state = "error"
                job.error = f"{type(e).__name__}: {e}"
            self._emit(job)


manager = JobManager()
